### DISCUSSION TITLE
Why does this revolve() fail?

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#23] [Q&A] [answered]
How to revolve a closed shape? I tried to follow the basic guide on the website. Here I learned that you can add thickness to a sketch with an extrude or revolve. I tried these methods in the following code, but I get a kernel error once I try to revolve a circle to create a donut. 

```` javascript
// addthickness experiments

/** @typedef { typeof import("replicad") } replicadLib */
/** @type {function(replicadLib, typeof defaultParams): any} */

function main( 
{
    Sketcher,
    sketchRectangle,
    sketchCircle,
})
{
let length = 60;
let width  = 15;
let height = 10;
let radius = 20;
let fillet = 4;

// create a sketch to be revolved
let sketch = new Sketcher("XZ")
    .hLine(width)
    .threePointsArc(0,2*fillet,fillet,fillet)
    .hLine(-width)
    .close()
let revolved = sketch.revolve()    

// create a sketch to be extruded
let rectangle = sketchRectangle(length,width,{plane: "XZ"})
let box = rectangle.extrude(30)

// create sketch of circle offset from x-axis and extrude this
let wheel = sketchCircle(radius,{plane:"XZ",origin: [30,0,0]}).extrude(2).translateY(-10)

// use same sketch and try to revolve this to create a donut => FAIL?? 
// let donut = sketchCircle(radius,{plane:"XZ",origin: [30,0,0]}).revolve()
 
box = box.fuse(wheel)
box = box.fuse(revolved)
// box = box.fuse(donut)

return box
}
````

**Comment by sgenoud** [comment, marked as answer]
The reason is that the revolve uses the origin of the sketch for its revolution axis. You need to specify another origin like that:

```js
 sketchCircle(20, {
    plane: "XZ",
    origin: [30, 0, 0],
  }).revolve([0, 0, 1], { origin: [0, 0, 0] });
```

It actually might be a better default to rotate on `[0, 0, 0]` always by default.

**Reply by evnp** [reply]
Just stumbled on this debugging the following code. Great that there's (almost) always an answer somewhere in discussions or resolved issue threads! I'd second this

> It actually might be a better default to rotate on [0, 0, 0] always by default.

it was unintuitive that various other drawings were working fine but `drawCircle` was failing (presumably because non-circle drawings _do_ use `[0, 0, 0]` as default origin).

```typescript
const pistonChannelTorus = drawCircle(pistonChannelDiam / 2)
  .sketchOnPlane(makePlane("XZ").translateX(pistonDiam / 2))
  .revolve([0, 0, 1])
```
Error:
```
StdFail_NotDone: BRep_API: command not done 
```
